File: app/star_pass_cli/_configuration.py
#!/usr/bin/env python3
""" What the deployment was configured with, and what it is missing.

    Its own module beside the two that show a run: '_render' shows
    what a run holds and '_sending' what would become of it, and a
    setting is neither.  The table and the labelled values come from
    '_render', because a table looks the same whichever answer it is
    showing.

    A calendar searched for nothing in particular carries one empty
    query string, which is a value and not an absence.  The contract
    publishes it as it is configured, so the wording that says what an
    empty one means is here, where wording belongs.

    The credential is shown beside the settings for the same reason it
    is read beside them: it is a fact about this deployment rather
    than about any run, and the four characters published of it are
    there to answer "which one is it running on".  The titles the data
    model has not matched are here on the same argument, and they are
    read by whoever is about to edit the model.
"""

# Imports - Python Standard Library
from typing import Any, Dict, List, Sequence

# Imports - Local
from ._render import labelled, NOTHING, table

# Constants
# What a calendar's row shows, in order.
CALENDAR_HEADERS = (
    'CALENDAR',
    'SEARCHED FOR'
)

# Said of a calendar whose window is read whole, in the column that
# would otherwise show what it is searched for.  A calendar configured
# with an empty query string searches for nothing in particular and
# therefore returns everything, so nothing in its window can be missed
# for want of a term -- which is what a reader of this column came to
# find out.
EVERYTHING = 'everything in the window'

# How the terms a calendar is searched for are joined.
TERM_GAP = ', '

# Headed rather than left to follow the values above it, because the
# two are different kinds of thing and a reader scanning for a
# calendar should not have to read the settings first.
CALENDARS_HEADING = 'CALENDARS'

# What a tested credential is said to be.  A whole word each way: the
# line is read at a glance and a tick and a cross are not one.
WORKING = 'working'
NOT_WORKING = 'not working'

# How the four published characters are shown.  Said as an ending
# rather than printed alone, so nobody reads them as the credential.
ENDING = 'ending {last_four}'

# What an unmatched title's row shows, in order.  The count is the
# column the reader is scanning: a title seen every month is a
# category the model is missing, and one seen once is an event that
# happened once.
UNMATCHED_HEADERS = (
    'CALENDAR',
    'TITLE',
    'SEEN',
    'LAST SEEN'
)

# Said when the log holds nothing, which is a finding rather than an
# empty screen: every title anybody recorded matched a category.
NOTHING_UNMATCHED = (
    'No unmatched titles have been recorded. Every title anybody has '
    'recorded matched a category.'
)


def searched_for(
        calendar: Dict[str, Any]
) -> str:
    """ Return what a calendar's window is searched for.

        Args:
            calendar (Dict[str, Any]):
                A calendar from an answer.

        Returns:
            text (str):
                The terms, or what an empty one means.
    """

    terms = [term for term in calendar['searchTerms'] if term]

    if not terms:
        return EVERYTHING

    return TERM_GAP.join(terms)


def calendar_row(
        calendar: Dict[str, Any]
) -> List[str]:
    """ Return one calendar as a row.

        Args:
            calendar (Dict[str, Any]):
                A calendar from an answer.

        Returns:
            row (List[str]):
                One value per column in 'CALENDAR_HEADERS'.
    """

    return [
        calendar['key'],
        searched_for(calendar=calendar)
    ]


def excluded_text(
        terms: Sequence[str]
) -> str:
    """ Return the terms a title is never collected under.

        Args:
            terms (Sequence[str]):
                The terms, as the contract publishes them.

        Returns:
            text (str):
                The terms, or a dash when the deployment excludes
                none.
    """

    return TERM_GAP.join(terms) or NOTHING


def config_text(
        answer: Dict[str, Any]
) -> str:
    """ Return the settings the service is running on.

        Args:
            answer (Dict[str, Any]):
                The configuration a client answered with.

        Returns:
            text (str):
                The settings, and the calendars under a heading of
                their own.
    """

    return '\n\n'.join(
        (
            labelled(
                pairs=(
                    ('Timezone', answer['timezone']),
                    ('Match threshold', str(answer['matchThreshold'])),
                    (
                        'Never collected',
                        excluded_text(terms=answer['excludedTitleTerms'])
                    )
                )
            ),
            CALENDARS_HEADING,
            table(
                headers=CALENDAR_HEADERS,
                rows=[
                    calendar_row(calendar=calendar)
                    for calendar in answer['calendars']
                ]
            )
        )
    )


def credential_text(
        answer: Dict[str, Any]
) -> str:
    """ Return whether the Amplify credential works.

        Args:
            answer (Dict[str, Any]):
                What a credential test answered.

        Returns:
            text (str):
                Whether it works and which one it is, with the reason
                below when it does not.
    """

    lines = labelled(
        pairs=(
            (
                'Amplify credential',
                WORKING if answer['working'] else NOT_WORKING
            ),
            (
                'Ends with',
                answer['lastFour'] or NOTHING
            )
        )
    )

    if answer['reason'] is None:
        return lines

    return f'{lines}\n\n{answer["reason"]}'


def unmatched_row(
        unmatched: Dict[str, Any]
) -> List[str]:
    """ Return one unmatched title as a row.

        Args:
            unmatched (Dict[str, Any]):
                An entry from an answer.

        Returns:
            row (List[str]):
                One value per column in 'UNMATCHED_HEADERS'.
    """

    return [
        unmatched['calendar'],
        unmatched['title'],
        str(unmatched['timesSeen']),
        unmatched['lastSeen']
    ]


def unmatched_text(
        answer: Sequence[Dict[str, Any]]
) -> str:
    """ Return the titles the data model has not matched.

        Args:
            answer (Sequence[Dict[str, Any]]):
                The entries a client answered with.

        Returns:
            text (str):
                A table, or a sentence when nothing has been recorded.
    """

    if not answer:
        return NOTHING_UNMATCHED

    return table(
        headers=UNMATCHED_HEADERS,
        rows=[
            unmatched_row(unmatched=unmatched)
            for unmatched in answer
        ]
    )

File: web/js/settings/configuration.js
/* What the service resolved at start up, and how long it keeps what a
 * run leaves behind.
 *
 * A pure reading of `GET /v1/config`: every value was answered by the
 * service, and nothing here works one out.
 *
 * **Two columns, not three.**  Where a value came from - a default,
 * the environment, fixed in the tool - is not published, so a third
 * column would be blank or invented.
 *
 * The words are this client's.  The contract answers with a zone, a
 * threshold and three numbers of days, and saying what each is
 * measured from is the job of whoever shows it.
 */

import { el, icon } from '../dom.js';

/* What a calendar searched for nothing in particular is said to be
 * searched for.  An empty query string is a value rather than an
 * absence: it returns the whole window, which is what makes a search
 * miss impossible on a calendar configured with one. */
const EVERYTHING = 'All events';

/* Stood in for a list the deployment configured nothing in. */
const NOTHING = '—';

/* How a list of terms is joined. */
const TERM_GAP = ', ';

/* The note under the settings table.  The zone above is the one a
 * window's dates are read in, and it belongs to the service rather
 * than to whoever is looking at this page. */
const TIMEZONE_NOTE = (
  'Runs read and display calendar data in the specified time zone, '
  + 'never the time zone of the browser.'
);

/* The note under the calendars table. */
const SEARCH_NOTE = (
  'Each term requires a separate search request to the League '
  + 'Calendar. Events that do not match a search term are classified '
  + 'as Not collected, and they may be added to a run manually.'
);

/* What each row of the settings table is called and how its value is
 * worded.  A table rather than three builders, because the difference
 * between the rows is data. */
const SETTINGS = [
  {
    name: 'Time zone',
    value: (config) => config.timezone
  },
  {
    name: 'Match confidence threshold',
    value: (config) => `${config.matchThreshold} out of 100`
  },
  {
    name: 'Filtered event titles',
    value: (config) => config.excludedTitleTerms.join(TERM_GAP) || NOTHING
  }
];

/* What is kept, for how long, and why that is the window it is on.
 * Three axes rather than one: the first two expire by age and
 * the third is answered by the data model coming to match a title,
 * and a single window over all three would delete the thing one of
 * them measures. */
const STORES = [
  {
    name: 'Job logs',
    kept: (retention) => `${retention.jobLogDays} days after the job finished`,
    note: 'The job itself is not removed with its log. The record of '
      + 'when a run occurred and its outcome survives longer than the '
      + 'run data.'
  },
  {
    name: 'Revisions',
    kept: (retention) => `${retention.revisionDays} days after the run `
      + 'last experienced activity',
    note: 'The first and current revisions persist indefinitely. '
      + 'Intermediate revisions expire.'
  },
  {
    name: 'Unmatched titles',
    kept: (retention) => `${retention.unmatchedTitleDays} days after a `
      + 'title was last seen',
    note: 'Unmatched titles expire if they are not added to the '
      + 'shift data model.'
  },
  {
    name: 'Sent shifts',
    kept: () => 'Indefinite',
    note: 'Retaining shift data sent to Amplify supports preventing '
      + 'attempts to create duplicate shifts.'
  }
];

/** Return what one calendar's window is searched for.
 *
 * @param {Object} calendar A calendar, as the configuration lists it.
 * @returns {string} The terms, or what an empty one means.
 */
function searchedFor(calendar) {
  const terms = calendar.searchTerms.filter((term) => term !== '');

  if (terms.length === 0) {
    return EVERYTHING;
  }

  return terms.map((term) => `"${term}"`).join(TERM_GAP);
}

/** Return a two-column table of names and values.
 *
 * @param {string} heading What the first column is called.
 * @param {string} second What the second column is called.
 * @param {Array<Array<string>>} rows The pairs, in order.
 * @returns {HTMLElement} The table.
 */
function pairTable(heading, second, rows) {
  return el(
    'table',
    /* The preview's table, which is the one table this interface
     * has. A second one styled separately would be two things that
     * have to look identical and no reason for them to stay that
     * way; what `settings-table` adds is the column widths, because
     * a name and a value are not a preview's five columns. */
    { class: 'table settings-table' },
    el(
      'thead',
      {},
      el(
        'tr',
        {},
        el('th', { text: heading }),
        el('th', { text: second })
      )
    ),
    el(
      'tbody',
      {},
      rows.map(([name, value]) => el(
        'tr',
        {},
        el('td', { text: name }),
        el('td', { class: 'mono settings-value', text: value })
      ))
    )
  );
}

/** Return one line under a table, saying what it means.
 *
 * @param {string} words The sentence.
 * @returns {HTMLElement} The line.
 */
function tableNote(words) {
  return el(
    'p',
    { class: 'settings-table-note muted note' },
    icon('info'),
    el('span', { text: words })
  );
}

/** Return the settings this service resolved.
 *
 * @param {Object} config What the deployment was configured with.
 * @returns {Array<Node>} The two tables and what they mean.
 */
export function resolvedSettings(config) {
  return [
    pairTable(
      'Setting',
      'Value',
      SETTINGS.map((setting) => [setting.name, setting.value(config)])
    ),
    tableNote(TIMEZONE_NOTE),
    el('h3', { class: 'settings-subheading', text: 'Calendars' }),
    pairTable(
      'Calendar',
      'Search terms',
      config.calendars.map(
        (calendar) => [calendar.key, searchedFor(calendar)]
      )
    ),
    tableNote(SEARCH_NOTE)
  ];
}

/** Return what is kept and for how long.
 *
 * @param {Object} retention The windows, as the configuration
 *     publishes them.
 * @returns {HTMLElement} The rows.
 */
export function stateRows(retention) {
  return el(
    'div',
    { class: 'settings-stores' },
    STORES.map((store) => el(
      'div',
      { class: 'row-card settings-store' },
      el(
        'div',
        {},
        el('div', { class: 'settings-store-name', text: store.name }),
        el('span', {
          class: 'mono muted note',
          text: store.kept(retention)
        })
      ),
      el('span', { class: 'muted meta settings-store-note', text: store.note })
    ))
  );
}
